src/main/daml/Daml/Finance/Instrument/Swap/Fpml/Util.daml

```diff
@@ -329,101 +343,196 @@ roundRate rate rounding =
         Down -> RoundingFloor
         Nearest -> RoundingHalfUp
 
+-- | HIDE
+-- Get the floating rate to be used for a period, taking stub periods into account.
+getFloatingRate : SwapStream -> SchedulePeriod -> SchedulePeriod ->
+  O -> HolidayCalendarData -> Bool -> O
+getFloatingRate s calculationPeriod paymentPeriod
+  regularRate rateFixingCalendar fixLeg =
+  case (calculationPeriod.stubType, s.stubCalculationPeriodAmount) of
+    (Some stubType, Some scpa) -> fromOptional regularRate $
+      getStubRate
+        scpa
+        (stubType == LongInitial || stubType == ShortInitial)
+        paymentPeriod
+        rateFixingCalendar
+        s.calculationPeriodDates.calculationPeriodDatesAdjustments.businessDayConvention
+        fixLeg
+    _ -> regularRate
+
+-- | HIDE
+-- Recursive function to calculate flat compounding.
+-- It follows the ISDA calculation example for flat compounding, example 4.2:
+-- https://www.isda.org/a/2KiDE/isda-compounding-memo.pdf
+calcFlatCompounding : O -> Deliverable -> SchedulePeriod -> SwapStream
+  -> O -> O -> HolidayCalendarData -> Bool ->
+  [(SchedulePeriod, Date, O)] ->  [(SchedulePeriod, Date, O)] -> [C]
+calcFlatCompounding notional currency p s
+  regularRateInclSpread regularRateExclSpread rateFixingCalendar fixLeg
+  (x::xs) calculationFixingsSelected =
+  calcFlatCompounding notional currency p s
+    regularRateInclSpread regularRateExclSpread rateFixingCalendar fixLeg
+    xs calculationFixingsSelected ++
+  calcFlatCompounding notional currency p s
+  regularRateInclSpread regularRateExclSpread rateFixingCalendar fixLeg
+  xs (calculationFixingsSelected ++ [x])
+calcFlatCompounding notional currency p s
+  regularRateInclSpread regularRateExclSpread rateFixingCalendar fixLeg
+  [] calculationFixingsSelected  =
+  if null calculationFixingsSelected then [] else [scale notional $ compoundedRates]
+    where
+    foldRates acc ((c, f, dcf), includeSpread) =
+      when (TimeGte f) $ scale (rate * dcf) acc
+      where
+        regularRate = if includeSpread then regularRateInclSpread else regularRateExclSpread
+        rate = getFloatingRate s c p regularRate rateFixingCalendar fixLeg
+
+    includeSpread = True :: replicate (length calculationFixingsSelected - 1) False
+
+    compoundedRates = foldl
+      foldRates
+      (when (TimeGte p.adjustedEndDate) $ one currency)
+      (reverse $ zip calculationFixingsSelected includeSpread)
+
+-- | HIDE
+-- Calculate notional and fixing data for a payment period.
+calculateNotionalAndFixings : SchedulePeriod -> SwapStream ->
```

Comment:
Thanks, the signature looks much better. Shall we consider introducing a record type also for the output? 

Also, I see other parts of the code following a similar pattern. Any change that `CalculationPeriodStaticData` can be used in there as well, to make it easier to read?

Reply:
Using a record type for the output would also make the signature of `calcFlatCompounding` more intelligible